stan/math/rev/mat/fun/gp_exponential_cov.hpp

```diff
@@ -0,0 +1,322 @@
+#ifndef STAN_MATH_REV_MAT_FUN_GP_EXPONENTIAL_COV_HPP
+#define STAN_MATH_REV_MAT_FUN_GP_EXPONENTIAL_COV_HPP
+
+#include <boost/math/tools/promotion.hpp>
+#include <boost/type_traits.hpp>
+#include <boost/utility/enable_if.hpp>
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/scal/fun/exp.hpp>
+#include <stan/math/prim/scal/fun/square.hpp>
+#include <stan/math/prim/scal/fun/squared_distance.hpp>
+#include <stan/math/prim/scal/err/check_not_nan.hpp>
+#include <stan/math/prim/scal/err/check_positive.hpp>
+#include <stan/math/prim/scal/meta/scalar_type.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/rev/scal/fun/value_of.hpp>
+#include <vector>
+
+namespace stan {
+namespace math {
+/**
+ * This is a subclass of the vari class for precomputed
+ * gradients of gp_exponential_cov
+ *
+ * The class stores the double values for the distance
+ * matrix, pointers to the varis for the covariance
+ * matrix, along with a pointer to the vari for sigma,
+ * and the vari for l.
+ *
+ * @tparam T_x type of std::vector of elements
+ * @tparam T_sigma type of sigma
+ * @tparam T_l type of length scale
+ */
+template <typename T_x, typename T_s, typename T_l>
```

Comment:
Questions:
1) any reason this is templated? It looks like the code only has one possible rev instantiation: `<double, var, var>`.
2) what's with the other two specializations? Just based on inspecting the code, it looks like there's going to be trouble with template argument deduction. It's got to be ambiguous the way it's written: `<T_x, double, T_l>` and `<T_x, T_s, double>`. I guess I'm saying I don't believe that it'll compile if you actually tried to have both in there.

Reply:
There are two different partial template specializations in the `rev` version of `cov_exp_quad`. 

L215 has: `cov_exp_quad<double, var, var>`
L257 has: `cov_exp_quad<double, double, var>`

That's why there is a base template function and then a partial specialization of `class cov_exp_quad_vari`. Now that I look at it, there's probably a different way to implement it, but that's a different question.